ath79: add support for TP-Link Archer C2 V3 (AC900)

x-logLT opened the pull request:
This commit adds support for TP-link Archer C2 V3 (AC900)

CPU: QCA9563 750Mhz
Ram: 64MB (DDR2)
Flash: 8MB (SPI NOR)
Ethernet: 5x 10/100/1000
Wifi: QCA9563 bgn + QCA9887 an+ac
9x Leds, 2x buttons

Signed-off-by: Skirmantas Lauzikas <skirmantas.lauzikas@blackraven.lt>

x-logLT committed:
ath79: add support for TP-Link Archer C2 V3 (AC900)

This commit adds support for TP-link Archer C2 V3 (AC900)

CPU: QCA9563 750Mhz
Ram: 64MB (DDR2)
Flash: 8MB (SPI NOR)
Ethernet: 5x 10/100/1000
Wifi: QCA9563 bgn + QCA9887 an+ac
9x Leds, 2x buttons

Signed-off-by: Skirmantas Lauzikas <skirmantas.lauzikas@blackraven.lt>

blocktrron commented on target/linux/ath79/dts/qca9563_tplink_archer-c2-v3.dts:
```diff
@@ -0,0 +1,195 @@
+
+/ {
+	compatible = "tplink,archer-c2-v3", "qca,qca9563";
+	model = "TP-Link Archer C2 Version 3";
```
I think this  should be more in line with other model-names, so i would go for "TP-LINK Archer C2 v3"

x-logLT replied:
Its seems there are at least two types of device naming schemes in ath79 branch, the one I used (also used for Archer c7, wr1043, wr841, ...), and the one you suggested. Are there any official guidelines on how devices should be named?

blocktrron replied:
Indeed, their naming seems to differ. In ar71xx, all TP-Link models were named like i proposed.

I think this is the much more sensible way as TP-Link refers to their devices always as "vX" instead of "Version X". But maybe there are other opinions.

blocktrron commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -398,6 +398,48 @@ static struct device_info boards[] = {
+			{"factory-boot", 0x00000, 0x20000},
+			{"fs-uboot", 0x20000, 0x10000},
+			{"os-image", 0x30000, 0x200000},	/* Stock: base 0x30000 size 0x100000 */
+			{"file-system", 0x230000, 0x5A0000},	/* Stock: base 0x130000 size 0x6a0000 */
```
It would be great if this device is using dynamic partitioning, which tplink-safeloader supports since this commit: https://github.com/openwrt/openwrt/commit/408c54b02f6a6dd2c0ae30a7885a37c9ac4c1b53

x-logLT committed:
firmware-utils: tplink-safeloader: switch Archer C2 V3 to dynamic partitioning

Switched Archer C3 v3 definition to dynamic partitioning, as per
suggestion by David Bauer <mail@david-bauer.net>

Signed-off-by: Skirmantas Lauzikas <skirmantas.lauzikas@blackraven.lt>

lars18th commented:
Hi,

Will this patch accepted?
I'm waiting for AC900 (Archer C2 v3) support.

x-logLT committed:
firmware-utils: tplink-safeloader: Fix for Archer C2 V3 image generation
Fix for factory image generation, so that stock firmware will accept
openwrt upgrade. As per suggestion by viper_lasser from
forum.openwrt.org and Heinz from https://tplinkforum.pl/

Signed-off-by: Skirmantas Lauzikas <skirmantas.lauzikas@blackraven.lt>

lars18th commented:
Still open and not merged?

clementperondevialet commented:
@lars18th Why not compile it and add a "Test-by" tag ?

@x-logLT  I think you should rebase on the latest version

clementperon commented:
@lars18th why not compile it and add a Tested-by tag ?

@x-logLT I think should rebase and fix the conflict with "target/linux/ath79/image/generic-tp-link.mk"

lars18th commented:
Hi @clementperon ,

Conflict resolved in PR #1579 .

lars18th commented:
Please, discard (**close**) this PR and review the version 2 #1579 .

rmilecki commented on target/linux/ath79/dts/qca9563_tplink_archer-c2-v3.dts:
```diff
@@ -0,0 +1,195 @@
+				reg = <0x020000 0x10000>;
+			};
+			
+			partition@30000 {
```
Please add `compatible`, see
[OpenWrt-Devel] Specifying "firmware" partition format on all DTS targets
http://lists.infradead.org/pipermail/openwrt-devel/2018-November/014706.html
for details.

lars18th commented:
Hi @x-logLT ,

1. Please, include changes from my PR #1579 to fix the conflict.
2. Please, add the change suggested by @rmilecki about the "firmware" partition format.

If you can't, then I suggest to some other developer to help and continue with the work.
Regards.

rmilecki commented:
One more comment: it seems
```
default-state = "off";
```
is not needed.

See https://github.com/openwrt/openwrt/commit/0d23fd2ab29a66f5d03187db4fac3e396b4f3b62

lars18th commented:
> One more comment: it seems
> 
> ```
> default-state = "off";
> ```
> is not needed.
> 
> See [0d23fd2](https://github.com/openwrt/openwrt/commit/0d23fd2ab29a66f5d03187db4fac3e396b4f3b62)

Sure!
And as I commented in https://github.com/openwrt/openwrt/pull/1579#issuecomment-448162792 I can't sign the patch. So I request that someone will create a new version of the patch with all conflicts resolved and merge it.

mkresin commented:
The install instructions form the commit message. The instructions are crucial and listed as requirement in the submitting patches guideline.

Everything else would be possible to fix while merging. But everything that needs to be fixed during merging may boggles us, as it means additional work for us.

yoksi commented:
hello, so is there someone who can push thru and finish this... "bureaucracy"?

jow- commented:
> hello, so is there someone who can push thru and finish this... "bureaucracy"?

I could help with solving the formal issues when someone provides an installation description.

lars18th commented:
> I could help with solving the formal issues when someone provides an installation description.

Hi @jow- ,
Thank you for your help!

Here you can found the description of how to install the OpenWRT firmware:
https://forum.openwrt.org/t/support-for-tp-link-archer-c2-v3/15617/26

And here how to rever to stock firmware:
https://forum.openwrt.org/t/support-for-tp-link-archer-c2-v3/15617/28

As you can see on this forum, at time only custom compilations can work. To provide "official" support only this patch (with the two simple conflicts resolved) is required. The "factory.bin" compatible firmware for first time install is provided by the regular OpenWRT toolchain, so no special guide is required.

Regards.

yoksi commented:
I wonder how many devices ended in similar situation without official dev support thanks to "last step" which no one can do.

Ferenc- commented:
> > hello, so is there someone who can push thru and finish this... "bureaucracy"?
> 
> I could help with solving the formal issues when someone provides an installation description.

Thanks @jow- for stepping up, please make this happen, I think many people are waiting for this device to be supported.

chunkeey commented:
Merged in https://github.com/openwrt/openwrt/commit/9c01273e5e7b3929ad46d974685ec29de207a8de

adrianschmutzler commented:
Although this PR has been has been closed already, I would be interested in some additional data about the device(s) in case you still have access to it:

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.
